lucene/misc/src/test/org/apache/lucene/misc/index/TestIndexRearranger.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.search.Sort;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.FixedBitSet;
+import org.apache.lucene.util.LuceneTestCase;
+
+public class TestIndexRearranger extends LuceneTestCase {
+  public void testRearrange() throws Exception {
+    Directory inputDir = newDirectory();
+    IndexWriter w =
+        new IndexWriter(
+            inputDir,
+            new IndexWriterConfig(null)
+                .setOpenMode(IndexWriterConfig.OpenMode.CREATE)
+                .setMergePolicy(NoMergePolicy.INSTANCE)
+                .setIndexSort(new Sort(new SortField("ord", SortField.Type.INT))));
+    for (int i = 0; i < 100; i++) {
+      Document doc = new Document();
+      doc.add(new NumericDocValuesField("ord", i));
+      w.addDocument(doc);
+      if (i % 10 == 9) {
+        w.commit();
+      }
+    }
+    IndexReader reader = DirectoryReader.open(w);
+    assertEquals(10, reader.leaves().size());
+    reader.close();
+    w.close();
+
+    Directory outputDir = newDirectory();
+    IndexRearranger rearranger =
+        new IndexRearranger(
+            inputDir,
+            outputDir,
+            new IndexWriterConfig(null)
+                .setOpenMode(IndexWriterConfig.OpenMode.CREATE)
+                .setMergePolicy(NoMergePolicy.INSTANCE)
+                .setIndexSort(new Sort(new SortField("ord", SortField.Type.INT))),
+            List.of(new OddDocSelector(), new EvenDocSelector()));
+    rearranger.execute();
+    reader = DirectoryReader.open(outputDir);
+    assertEquals(2, reader.leaves().size());
+    LeafReader segment1 = reader.leaves().get(0).reader();
+    assertEquals(50, segment1.numDocs());
+    NumericDocValues numericDocValues = segment1.getNumericDocValues("ord");
+    assertTrue(numericDocValues.advanceExact(0));
+    boolean odd = numericDocValues.longValue() % 2 == 1;
+    for (int i = 1; i < 50; i++) {
+      assertTrue(numericDocValues.advanceExact(i));
+      if (odd) {
+        assertEquals(1, numericDocValues.longValue() % 2);
+      } else {
+        assertEquals(0, numericDocValues.longValue() % 2);
+      }
+    }
+    LeafReader segment2 = reader.leaves().get(0).reader();
+    assertEquals(50, segment2.numDocs());
+    numericDocValues = segment2.getNumericDocValues("ord");
+    assertTrue(numericDocValues.advanceExact(0));
+    odd = numericDocValues.longValue() % 2 == 1;
+    for (int i = 1; i < 50; i++) {
+      assertTrue(numericDocValues.advanceExact(i));
+      if (odd) {
+        assertEquals(1, numericDocValues.longValue() % 2);
+      } else {
+        assertEquals(0, numericDocValues.longValue() % 2);
+      }
+    }
+    reader.close();
+    inputDir.close();
+    outputDir.close();
+  }
+
+  private class OddDocSelector implements IndexRearranger.DocumentSelector {
```

Comment:
This looks very similar to "shard splitting" implementations!

I think an interesting `DocumentSelector` might be one which assigns documents into a "perfect" logarithmic staircase according to some base and exponent, e.g. target 7 segments with 1M docs, 7 segments with 100K docs, 7 segments with 10K docs.  This would allow us to take any index and rearrange it into arbitrary multi-segment situations.  Fun!

Reply:
Yeah, in extreme case we might be able to optimize the search performance by changing document distribution